### PR TITLE
watch + services

### PR DESCRIPTION
**Pull request by davidhadas**
PR includes: 
1. cmd/guard-service/service.go that implements a service cache class  - maintaining a record per deployed service 
    All per-service actions of the guard-service are implemented by this class
2. Add Watch to KubeMgr

### CODE REVIEW

**Comment by davidhadas** [comment]
I will need to add at least part of guard-service/main.go  - as it is where `var log` is defined and initialized.

**Comment by davidhadas** [comment]
Better to leave code coverage to the next PR, cmd/guard-service/main.go should be better covered by then.

File: cmd/guard-service/main.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	GuardServicePort     string `split_words:"true" required:"false"`
+}
+
+type learner struct {
+	services        *services
+	pileLearnTicker utils.Ticker
+}
+
+func (l *learner) fetchConfig(w http.ResponseWriter, req *http.Request) {
+}
```
Comment (maximilien) [inline cmd/guard-service/main.go:43]: Why is this function empty?

File: cmd/guard-service/main.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+func (l *learner) processPile(w http.ResponseWriter, req *http.Request) {
+	w.Write([]byte{})
+}
+
+func (l *learner) mainEventLoop(quit chan int) {
+
```
Comment (maximilien) [inline cmd/guard-service/main.go:50]: New line?

File: cmd/guard-service/main.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	for {
+		select {
+		case <-l.pileLearnTicker.Ch():
+			l.services.tick()
+		case <-quit:
+			log.Info("mainEventLoop was asked to quit!")
```
Comment (maximilien) [inline cmd/guard-service/main.go:56]: Can we include some reason as we quit?

File: pkg/guard-kubemgr/watcher.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	"fmt"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	spec "knative.dev/security-guard/pkg/apis/guard/v1alpha1"
+)
+
```
Comment (maximilien) [inline pkg/guard-kubemgr/watcher.go:31]: Would it make sense to create a Watcher interface?

**Review by maximilien — COMMENTED** [review]
Left a few feedback.

Also might help to have some diagram description of the key interfaces and their interactions with the rest of the Knative objects?

File: cmd/guard-service/main.go (hunk 2 above)
Comment (davidhadas) [reply cmd/guard-service/main.go:50]: can be removed...

File: pkg/guard-kubemgr/watcher.go (hunk 4 above)
Comment (davidhadas) [reply pkg/guard-kubemgr/watcher.go:31]: I did not identify a reason to do that. Was not needed so far. 
KubeMgr has an interface with a watch method

File: cmd/guard-service/main.go (hunk 3 above)
Comment (davidhadas) [reply cmd/guard-service/main.go]: I added a quit reason

**Comment by davidhadas** [comment]
> Also might help to have some diagram description of the key interfaces and their interactions with the rest of the Knative objects?

@maximilien see https://github.com/knative-sandbox/security-guard/blob/main/ARCHITECTURE.md and advise what needs to be added

**Comment by davidhadas** [comment]
The coverage report seems wrong... shows main is 0% tested after adding a test

**Review by maximilien — APPROVED** [review]
/approve
/lgtm

**Review by maximilien — APPROVED** [review]
/approve
/lgtm

**Pull request merged**